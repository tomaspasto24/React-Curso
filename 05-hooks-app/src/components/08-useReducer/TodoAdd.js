import React from 'react'
import { useForm } from '../../hooks/useForm';

export const TodoAdd = ( { handleAddTodo } ) => {
    
    const [ { description }, handleInputChange, reset ] = useForm({
        description: "",
    });

    const handleSubmit = (e) => {
        e.preventDefault();

        if ( description.trim().length <= 2 ){
            return;
        };

        const newTodo = {
            id: new Date().getTime(),
            desc: description,
            done: false
        }

        handleAddTodo(newTodo);
        reset();
    }

    return (
        <>
            <h4>Agregar TODO</h4>
            <hr />

            <form onSubmit={ handleSubmit }>
                <input 
                    type='text'
                    name='description'
                    placeholder='Agregar tarea...'
                    className='form-control'
                    value={ description }
                    onChange={ handleInputChange }
                />

                <div className = "d-grid gap-2">
                    <button 
                        type='submit'
                        className='btn btn-outline-primary mt-1'
                    >
                        Agregar
                    </button>
                </div>

            </form>  
        </>
    )
}
