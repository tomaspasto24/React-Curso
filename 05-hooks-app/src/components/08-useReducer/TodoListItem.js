import React from 'react'
import PropTypes from 'prop-types';

export const TodoListItem = ( { todo, index, handleDelete, handleToggle } ) => {
    return (
        <li
                key={todo.id}
                className='list-group-item'
            >

                <p
                    onClick={ () => handleToggle(todo.id) } 
                    className={ `${todo.done && 'complete'}` }
                >
                    { index+1 }. { todo.desc }
                </p>

                <button
                    onClick={ () => handleDelete(todo.id) } 
                    className='btn btn-danger'
                >
                    Borrar
                </button>

        </li>
    )
}

TodoListItem.propTypes = {
    todo: PropTypes.object.isRequired,
    index: PropTypes.number.isRequired,
    handleDelete: PropTypes.func.isRequired,
    handleToggle: PropTypes.func.isRequired,
}