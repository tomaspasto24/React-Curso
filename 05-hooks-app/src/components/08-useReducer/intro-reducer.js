
const initialState = [{
    id:1,
    todo: 'Comprar pan',
    done: false
}];

const todoReducer = ( state = initialState, action ) => {

    if (action?.type === 'agregar'){
        state = [ ...state, action.payload ]
    }

    return state;
};


let todos = todoReducer();

const newTodo = [{
    id:2,
    todo: 'salir a correr',
    done: false
}]

const action = {
    type: 'agregar',
    payload: newTodo
}

todos = todoReducer(todos, action);

console.log(todos);