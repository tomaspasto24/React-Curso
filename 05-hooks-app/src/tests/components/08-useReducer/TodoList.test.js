import React from 'react';
import { shallow } from 'enzyme';
import { TodoList } from '../../../components/08-useReducer/TodoList';
import { demoTodos } from '../../fixtures/demoTodos';

describe('Pruebas en TodoList.', () => {

    const handleDelete = jest.fn();
    const handleToggle = jest.fn();

    const wrapper = shallow( <TodoList
        todos={demoTodos}
        handleDelete={ handleDelete }
        handleToggle={ handleToggle }
        /> );

    test('Debe de mostrarse correctamente.', () => {

        expect( wrapper ).toMatchSnapshot();
    });

    test('Debe tener dos TodoListItem.', () => {

        expect( wrapper.find('TodoListItem').length ).toBe( demoTodos.length );
        //espera que la propiedad sea cualquier funci√≥n 
        expect( wrapper.find('TodoListItem').at(0).prop('handleDelete') ).toEqual( expect.any(Function) );
    });

});