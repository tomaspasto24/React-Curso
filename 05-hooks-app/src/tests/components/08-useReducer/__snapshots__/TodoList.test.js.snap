// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en TodoList. Debe de mostrarse correctamente. 1`] = `
<Fragment>
  <ul
    className="list-group list-grup-flush"
  >
    <TodoListItem
      handleDelete={[MockFunction]}
      handleToggle={[MockFunction]}
      index={0}
      key="2"
      todo={
        Object {
          "desc": "Aprender react",
          "done": false,
          "id": 2,
        }
      }
    />
    <TodoListItem
      handleDelete={[MockFunction]}
      handleToggle={[MockFunction]}
      index={1}
      key="1"
      todo={
        Object {
          "desc": "Aprender mongo",
          "done": false,
          "id": 1,
        }
      }
    />
  </ul>
</Fragment>
`;
